signac/core/buffered_collection.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implement the buffering feature to  SyncedCollection API."""
+import os
+import errno
+import logging
+from contextlib import contextmanager
+from abc import abstractmethod
+
+from .synced_collection import SyncedCollection
+from .synced_list import SyncedList
+from .syncedattrdict import SyncedAttrDict
+from .errors import Error
+
+logger = logging.getLogger(__name__)
+
+_BUFFERED_MODE = 0
+_BUFFERED_MODE_FORCE_WRITE = None
+_BUFFER = dict()
+_FILEMETA = dict()
+
+
+class BufferException(Error):
+    """An exception occured in buffered mode."""
+
+    pass
+
+
+class BufferedFileError(BufferException):
+    """Raised when an error occured while flushing one or more buffered files.
+
+    .. attribute:: files
+
+        A dictionary of files that caused issues during the flush operation,
+        mapped to a possible reason for the issue or None in case that it
+        cannot be determined.
+    """
+
+    def __init__(self, files):
+        self.files = files
+
+    def __str__(self):
+        return "{}({})".format(type(self).__name__, self.files)
+
+
+def _get_filemetadata(filename):
+    try:
+        return os.path.getsize(filename), os.path.getmtime(filename)
+    except OSError as error:
+        if error.errno != errno.ENOENT:
+            raise
+
+
+def _store_in_buffer(id, backend_kwargs, cache, metadata=None):
+    assert _BUFFERED_MODE > 0
+    _BUFFER[id] = (backend_kwargs, cache)
+    if not _BUFFERED_MODE_FORCE_WRITE:
+        _FILEMETA[id] = metadata
+
+
+def flush_all():
+    """Execute all deferred JSONDict write operations."""
+    logger.debug("Flushing buffer...")
+    issues = dict()
+    while _BUFFER:
+        id, (backend_kwargs, cache) = _BUFFER.popitem()
+        if not _BUFFERED_MODE_FORCE_WRITE:
+            meta = _FILEMETA.pop(id)
+            # extract the data from cache
+            backend_class = SyncedCollection.from_backend(backend_kwargs['backend'])
+            data = backend_class._read_from_cache(id, cache)
+
+            obj = SyncedCollection.from_base(data, cache=cache, no_sync=True, **backend_kwargs)
+            if obj._get_metadata() != meta:
+                issues[id] = 'File appears to have been externally modified.'
+                continue
+            try:
+                obj._sync()
+            except OSError as error:
+                logger.error(str(error))
+                issues[id] = error
+    if issues:
+        raise BufferedFileError(issues)
+
+
+def in_buffered_mode():
```

Comment:
This method is from the previous implementation of buffering. I have tried to match the API as much as possible.

Reply:
@vishav1771 no need to necessarily match old APIs. If checking whether something is buffered is something that signac has to do, then we can leave this public, but I think based on our more recent discussions it should be private since whether or not we're in buffered mode should be entirely transparent to signac.